rotkehlchen/exchanges/bitstamp.py

```diff
@@ -0,0 +1,711 @@
+        """Given a user transaction that contains the base and quote assets'
+        symbol as keys, return the Bitstamp trade pair data (raw pair str,
+        base/quote assets raw symbols, and TradePair).
+
+        NB: any custom pair conversion (e.g. from Bitstamp asset symbol to world)
+        should happen here.
+
+        Can raise DeserializationError.
+        """
+        try:
+            pair = [key for key in raw_result.keys() if '_' in key and key != 'order_id'][0]
+        except IndexError:
+            raise DeserializationError(
+                'Could not deserialize Bitstamp trade pair from user transaction. '
+                f'Trade pair not found in: {raw_result}.',
+            )
+
+        # NB: `pair_get_assets()` is not used for simplifying the calls and
+        # storing the raw pair strings.
+        base_asset_symbol, quote_asset_symbol = pair.split('_')
+        try:
+            base_asset = Asset(base_asset_symbol)
+            quote_asset = Asset(quote_asset_symbol)
+        except UnknownAsset as e:
+            log.error(str(e))
+            raise DeserializationError(
+                f'Unknown asset {e.asset_name} found while processing trade pair.',
+            )
+
+        return TradePairData(
+            pair=pair,
+            base_asset_symbol=base_asset_symbol,
+            quote_asset_symbol=quote_asset_symbol,
+            base_asset=base_asset,
+            quote_asset=quote_asset,
+            trade_pair=TradePair(f'{base_asset.identifier}_{quote_asset.identifier}'),
+        )
+
+    @staticmethod
+    def _check_for_system_clock_not_synced_error(
+            error_code: Optional[int] = None,
+    ) -> None:
+        if error_code == API_SYSTEM_CLOCK_NOT_SYNCED_ERROR_CODE:
+            raise SystemClockNotSyncedError(
+                current_time=str(datetime.now()),
+                remote_server='Bitstamp',
+            )
+
+        return None
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['validate_api_key'],
+    ) -> Tuple[bool, str]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['balances'],
+    ) -> Tuple[Optional[Dict[Asset, Balance]], str]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['trades'],
+    ) -> List[Trade]:
+        ...
+
+    @overload  # noqa: F811
+    def _process_unsuccessful_response(  # pylint: disable=no-self-use
+            self,
+            response: Response,
+            case: Literal['asset_movements'],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _process_unsuccessful_response(
```

Comment:
@LefterisJP I've implemented this method for reducing the blocks related with processing unsuccessful requests, and standardize it a bit more (sometimes between exchanges differ in some logs up or down). Also cause it has really helped me to see the subtle differences between methods, and it has been much easier to debug and test.

Reply:
Sure no problem. It looks nice and feels neat to have this logic abstracted away in a function like that.